server/src/main/java/org/elasticsearch/cluster/project/ProjectStateRegistry.java

```diff
@@ -343,48 +363,113 @@ public ProjectStateRegistry build() {
         }
     }
 
-    private record Entry(Settings settings) implements Writeable, Diffable<Entry> {
+    private record Entry(Settings settings, ImmutableOpenMap<String, ReservedStateMetadata> reservedStateMetadata)
+        implements
+            ToXContentFragment,
+            Writeable,
+            Diffable<Entry> {
+
+        Entry() {
+            this(Settings.EMPTY, ImmutableOpenMap.of());
+        }
 
         public static Entry readFrom(StreamInput in) throws IOException {
-            return new Entry(Settings.readSettingsFromStream(in));
+            Settings settings = Settings.readSettingsFromStream(in);
+
+            ImmutableOpenMap<String, ReservedStateMetadata> reservedStateMetadata;
+            if (in.getTransportVersion().onOrAfter(TransportVersions.PROJECT_RESERVED_STATE_MOVE_TO_REGISTRY)) {
+                int reservedStateSize = in.readVInt();
+                ImmutableOpenMap.Builder<String, ReservedStateMetadata> builder = ImmutableOpenMap.builder(reservedStateSize);
+                for (int i = 0; i < reservedStateSize; i++) {
+                    ReservedStateMetadata r = ReservedStateMetadata.readFrom(in);
+                    builder.put(r.namespace(), r);
+                }
+                reservedStateMetadata = builder.build();
+            } else {
+                reservedStateMetadata = ImmutableOpenMap.of();
+            }
+
+            return new Entry(settings, reservedStateMetadata);
         }
 
         public Entry withSettings(Settings settings) {
-            return new Entry(settings);
+            return new Entry(settings, reservedStateMetadata);
+        }
+
+        public Entry withReservedStateMetadata(ReservedStateMetadata reservedStateMetadata) {
+            ImmutableOpenMap<String, ReservedStateMetadata> build = ImmutableOpenMap.builder(this.reservedStateMetadata)
```

Comment:
Nit: a better variable name?
```suggestion
            final var reservedStateMetadataMap = ImmutableOpenMap.builder(this.reservedStateMetadata)
```